### ISSUE TITLE
Package has some weird version on npm

### ISSUE THREAD

**Issue by brunokrebs** [Foxandxss/koa-unless#3]
I was having trouble executing a simple test because `koa` was complaining about a non `generator` function being returned by `koa-unless`. Then I came here to check the source code and I couldn't find any commit or version (actually there are no releases/tags) that were returning this simple function.

So, after a hard time, I finally realised that the problem is that the package on NPM has this version that does not return a `generator` function. I have no idea how it got there, but could you please update it, @Foxandxss?

This is how I tracked the issue:

```bash
npm view koa-unless dist.tarball
```

The command above shows the url that contains the code that is downloaded when issuing a `npm install koa-unless`. As you can see, [the tarball contains this index.js](https://registry.npmjs.org/koa-unless/-/koa-unless-1.0.0.tgz) that returns the simple function:

```js
module.exports = function(options) {
    return function(ctx, next) {
        ...
    }
}
```

**Comment by Foxandxss** [comment]
Sure. I can also give you powers if you want.

**Comment by brunokrebs** [comment]
Sounds nice @Foxandxss. Thanks

**Comment by brunokrebs** [comment]
@Foxandxss, could you add me as a maintainer of this package on npm? It is just a matter of issuing:

```bash
npm owner add brunokrebs
```

[Npm docs regarding adding users](https://docs.npmjs.com/cli/owner).

**Comment by alex-friedl** [comment]
Would be great if this issue could be resolved one or the other way! :)

**Comment by brunokrebs** [comment]
It is up to @Foxandxss. He can update it or he can add me as owner so I can update it myself.

**Comment by Foxandxss** [comment]
Didn't I gave you powers Bruno?

**Comment by brunokrebs** [comment]
Well, you have never answered my request. You gave me access to this repo, but I'm not aware if you gave me access to update the NPM module. Which is done with the following command:

```sh
npm owner add brunokrebs
```

I will try to update the module on NPM it this weekend.

**Comment by Foxandxss** [comment]
I added you the first day. Just checked

**Comment by brunokrebs** [comment]
Cool, thanks @Foxandxss. I do have access and I have just updated it on NPM. Now it is on version 1.0.1 and the issue is fixed. /cc @alex-friedl

**Comment by jozsi** [comment]
@brunokrebs - 1.0.1 broke my tests.

Until yesterday (V1.0.0) following was working:
```
router.use('/user', jwt.unless({ method: 'POST' }), user.routes());
```

since then (currently at v1.0.1), it returns a 404 error.

The git commit log also doesn't reflect the changes between the two versions, so I lost track what actually went wrong.

**Comment by sdd** [comment]
<img width="751" alt="screen shot 2017-03-18 at 13 58 57" src="https://cloud.githubusercontent.com/assets/323439/24072719/183945f0-0be3-11e7-8443-82aa6910f314.png">

Confirmed. Reproduced by checking out `koa-jwt` v2 branch, wiping out node_modules, `npm install`, `npm test`.

**Comment by onbjerg** [comment]
Same issue here as @jozsi. Pretty frustrating. Pinning to 1.0.0.

**Comment by brunokrebs** [comment]
I have published a new patch (1.0.3). Could you guys take a look? //cc @onbjerg @jozsi 

Thanks

ps. 1.0.2 is buggy

**Comment by silverzhaojr** [comment]
@brunokrebs 
Hi, in [line #22](https://github.com/Foxandxss/koa-unless/blob/1168a89f9e731eed4631a27b542d0d0daf6bcb46/index.js#L22) of file `index.js`:

```javascript
22  if (matchesCustom(this, opts) || matchesPath(requestedUrl, opts) ||
23      matchesExtension(requestedUrl, opts) || matchesMethod(ctx.method, opts)) {
24    return next();
25  }
```

I think `matchesCustom(this, opts)` should be `matchesCustom(ctx, opts)`, or in function `matchesCustom` it cannot get correct `ctx`.

Could you please confirm it? Thanks!

**Comment by nfantone** [comment]
@silver-zhao Came here exactly for this. That should definitely be `ctx`, rather than `this`. And [further below](https://github.com/Foxandxss/koa-unless/blob/1168a89f9e731eed4631a27b542d0d0daf6bcb46/index.js#L40), it shouldn't be using `.call`.

(cc. @brunokrebs)

**EDIT**: Unless, this is not targeting Koa `^2.0.0`?

**Comment by nfantone** [comment]
@brunokrebs Also, tests are failing. I'd be happy to put together a PR if you could instruct me on what path to follow. I can see `"devDependencies"` declares a very old version of Koa (0.21.0), but the actual middleware doesn't return a generator and is using Koa 2 syntax.

**Comment by brunokrebs** [comment]
@nfantone this is one of the problems. The version `1.0.0` released on NPM uses a code that is not available on this repo. Probably @Foxandxss developed it before creating the repo and then released it on NPM. And then, after that, he created the repo and the tests.

I'm working on this issue right now. I'll submit a PR and then you guys can take a look and make comments before releasing it again. Thanks

**Comment by jozsi** [comment]
I'd suggest to have a branch for Koa v2 (using `ctx`) and a legacy branch for older Koa (using `this`).

**Comment by nfantone** [comment]
@brunokrebs Could use this if you like: https://github.com/Foxandxss/koa-unless/pull/5

**Comment by nfantone** [comment]
@jozsi I'm not familiar with how releases were managed in this project, but I don't see the point in maintaining a legacy branch for older versions of Koa. You'd just need to install a compatible version of `koa-unless` for your project and properly document the version jump.

**Comment by jozsi** [comment]
Most of the Koa related modules have a major version bump for Koa v2.

Let's take `koa-jwt` as an example, as it contributes to over 90% of `koa-unless` downloads.

Their [Koa v1 branch](https://github.com/koajs/jwt/blob/master/package.json) is versioned v1.3.0 and uses `jwt-unless` v0.0.1.

Their [Koa v2 branch](https://github.com/koajs/jwt/blob/koa-v2/package.json) is versioned v3.0.1 and uses `jwt-unless` v1.0.0.

I would keep the Koa v1 compatible versions of `jwt-unless` as `0.0.x` (note: `0.1.0` would not satisfy semver `^0.0.1`).

For the Koa v2 compatible version, I would use v1 because most (if not all) of the packages that depend on `koa-unless` are using `koa v2` and `koa-unless v1`.

**Comment by brunokrebs** [comment]
Would you mind to take a look at PR [#6](https://github.com/Foxandxss/koa-unless/pull/6). I think I managed to make it compatible with Koa `0.X`, `1.X` and `2.X`. But I would like feedback before publish buggy versions as I did before.

If you guys find anything we may consider accepting the PR, then adding more tests to the test cases and then fix them.

//cc @jozsi @nfantone @silver-zhao @Foxandxss

**Comment by brunokrebs** [comment]
Well, I finally realize what heck has happened. The whole problem is that I didn't noted that `koa-unless@v0.0.1` was the code under `master` and that `koa-unless@v1.0.0` was the code under `koa-v2`.

I have now updated both `master` and `koa-v2` properly and published the correct versions on NPM. Now I will upport the refactoring that I have made on `master` to `koa-v2`, which shall improve readability.

Sorry about the mess.
